### ISSUE TITLE
Bad plugins break Girder 3

### ISSUE THREAD

**Issue by manthey** [girder/girder#2889]
In Girder 3, we need to guard against bad plugins.  That is, suppose I have a plugin whose `__init__.py` script contains:

```
import unknown_module        # This is the problem
from girder.plugin import GirderPlugin

class MyPlugin(GirderPlugin):
  ...
```

with an appropriate entrypoint in its setup.py.  When Girder calls `_getPluginRegistry`, the `entryPoint.load()` call throws an uncaught exception and Girder never starts.  These failures occur before the plugin can be added to the _pluginRegistry (since we never learn its internal name).

We need to catch and log this exception.

Should we also add an approximate value to plugin registry?  That is, should we do something like
```
for entryPoint in iter_entry_points(_NAMESPACE):
        try:
            pluginClass = entryPoint.load()
            plugin = pluginClass(entryPoint)
            _pluginRegistry[plugin.name] = plugin
        except Exception:
            try:
                plugin = GirderPlugin(entryPoint)
                plugin._pluginFailureInfo = sys.exc_info()
                _pluginRegistry[entryPoint.name] = plugin
                logprint.exception('Failed to instantiate plugin %s' % entryPoint.name)
            except Exception:
                logprint.exception('Failed to parse plugin %s' % entryPoint.name)
```

As a further minor bug, Girder 3 no longer shows the failure messages from failed plugins on the web client (an easy change to the pug file will fix it).

**Comment by jbeezley** [comment]
This throws an exception like this:
```python
Traceback (most recent call last):
  File "/home/jbeezley/git/girder3/env/bin/girder", line 11, in <module>
    load_entry_point('girder', 'console_scripts', 'girder')()
  File "/home/jbeezley/git/girder3/env/lib/python3.6/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/home/jbeezley/git/girder3/env/lib/python3.6/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/home/jbeezley/git/girder3/env/lib/python3.6/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/jbeezley/git/girder3/env/lib/python3.6/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/jbeezley/git/girder3/env/lib/python3.6/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/home/jbeezley/git/girder3/girder/cli/serve.py", line 44, in main
    server.setup(testing)
  File "/home/jbeezley/git/girder3/girder/utility/server.py", line 213, in setup
    girderWebroot, appconf = configureServer(test, plugins, curConfig)
  File "/home/jbeezley/git/girder3/girder/utility/server.py", line 160, in configureServer
    plugin._loadPlugins(plugins, info)
  File "/home/jbeezley/git/girder3/girder/plugin.py", line 238, in _loadPlugins
    pluginObject = getPlugin(name)
  File "/home/jbeezley/git/girder3/girder/plugin.py", line 216, in getPlugin
    registry = _getPluginRegistry()
  File "/home/jbeezley/git/girder3/girder/plugin.py", line 208, in _getPluginRegistry
    pluginClass = entryPoint.load()
  File "/home/jbeezley/git/girder3/env/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2343, in load
    return self.resolve()
  File "/home/jbeezley/git/girder3/env/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2349, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/home/jbeezley/git/girder3/env/lib/python3.6/site-packages/girder_worker/girder_plugin/__init__.py", line 23, in <module>
    from girder_jobs.models.job import Job
  File "/home/jbeezley/git/girder3/plugins/jobs/girder_jobs/__init__.py", line 20, in <module>
    import invalid_module
ModuleNotFoundError: No module named 'invalid_module'
```

It tells you where it is failing and why.  To me, that is the correct behavior.  If you have a broken plugin installed, girder shouldn't start.

**Comment by manthey** [comment]
What is our logic for starting with a broken plugin where the `load` method is missing but not when a python module is missing?

**Comment by jbeezley** [comment]
I would make a distinction between errors that can be occur in an endpoint and one that requires direct access to the server.  I would be okay with making the lack of a load method an error at server start; however, you can't detect all possible problems with loading a plugin unless you actually load it.

**Comment by zachmullen** [comment]
I could see the argument for wanting Girder to be able to start up even in this case, maybe via a flag or something. In many cases, particularly in dev environments where one might be switching between versions of core, it is very possible to have broken plugins installed but with no intent to ever load them, and having that make Girder unable to start would be an annoyance.

**Comment by manthey** [comment]
I'm happy to make a PR that catches this.  I'm equally happy to make a PR that just fixes the web client reporting existing errors.

I'd rather not make it conditional on a flag.

**Comment by zachmullen** [comment]
If nobody is pushing strongly for the ignore-error behavior, let's just do the web client fix. Thanks @manthey

**Comment by manthey** [comment]
Closed via #2890.
